### PR TITLE
OPENEUROPA-3409: Pluggable subsets of concepts

### PR DESCRIPTION
**Pull request by upchuk**
This PR creates the `ConceptSubset` plugin type that allows the definition of a subset of concepts within which the SKOS Concept entity reference selection plugin can query.

Subsets are used to define adhoc groups of concepts within a scheme that can be configured and used in an en entity reference field that references SKOS Concepts.

### CODE REVIEW

File: src/Plugin/EntityReferenceSelection/SkosConceptSelection.php [hunk 1]
```diff
@@ -52,17 +95,26 @@ public function buildConfigurationForm(array $form, FormStateInterface $form_sta
...
       '#multiple' => TRUE,
+      '#size' => 10,
+      // Ajax is applied in
+      // EntityReferenceItem::fieldSettingsAjaxProcessElement()
+      '#ajax' => TRUE,
     ];
 
-    if (empty($concept_scheme_ids)) {
-      return $form;
+    $concept_schemes = $configuration['concept_schemes'] ?
```
[the remark this answers is not in the document]
Comment (upchuk) [reply src/Plugin/EntityReferenceSelection/SkosConceptSelection.php:104]: true

File: src/Plugin/EntityReferenceSelection/SkosConceptSelection.php [hunk 2]
```diff
@@ -112,12 +164,103 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
...
+   *   The value to match.
+   */
+  protected function applyConceptSubset(QueryInterface $query, $match_operator, array $concept_schemes = [], string $match = NULL) {
+    $configuration = $this->getConfiguration();
+    if (!$configuration['concept_subset']) {
+      return;
+    }
+
+    /** @var \Drupal\rdf_skos\ConceptSubsetInterface $plugin */
+    $plugin = $this->subsetManager->createInstance($configuration['concept_subset']);
```
[the remark this answers is not in the document]
Comment (upchuk) [reply src/Plugin/EntityReferenceSelection/SkosConceptSelection.php:262]: yeah, we do :)

File: src/ConceptSubsetPluginManager.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getPredicateMappingDefinitions(): array {
+    $definitions = $this->getDefinitions();
+    $predicate_mappers = [];
+    foreach ($definitions as $id => $definition) {
+      if (isset($definition['predicate_mapping']) && (bool) $definition['predicate_mapping']) {
```
[the remark this answers is not in the document]
Comment (upchuk) [reply src/ConceptSubsetPluginManager.php:46]: ah ok :)

File: src/Plugin/EntityReferenceSelection/SkosConceptSelection.php [hunk 4]
```diff
@@ -42,8 +87,6 @@ public function buildConfigurationForm(array $form, FormStateInterface $form_sta
     $form = parent::buildConfigurationForm($form, $form_state);
```
[the remark this answers is not in the document]
Comment (upchuk) [reply src/Plugin/EntityReferenceSelection/SkosConceptSelection.php:87]: @brummbar Removed from the form. removing the default values from the parents is a bit too complicated as we need some of it + the scheme would not match anymore with previous configs.

File: src/ConceptSubsetPluginManager.php [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+    if (!$plugin_ids) {
+      return [];
+    }
+
+    $plugin_ids = count($plugin_ids) === 1 ? reset($plugin_ids) : call_user_func_array('array_intersect', $plugin_ids);
+    $definitions = [];
+    foreach ($plugin_ids as $plugin_id) {
+      $definitions[$plugin_id] = $this->getDefinition($plugin_id);
+    }
```
Comment (brummbar) [inline src/ConceptSubsetPluginManager.php:73]:
```suggestion
    $definitions_by_scheme = [];
    foreach ($concept_schemes as $concept_scheme) {
      $definitions_by_scheme[$concept_scheme] = $this->getDefinitionsForConceptScheme($concept_scheme);
    }

    if (!$definitions_by_scheme) {
      return [];
    }

    return count($definitions_by_scheme) === 1 ? reset($definitions_by_scheme) : call_user_func_array('array_intersect_key', $definitions_by_scheme);
```

**Pull request merged**